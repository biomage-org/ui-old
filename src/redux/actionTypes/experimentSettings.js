const EXPERIMENT_SETTINGS = 'experimentSettings';

/**
 * Loaded general information about the experiment.
 */
const EXPERIMENT_SETTINGS_INFO_UPDATE = `${EXPERIMENT_SETTINGS}/updateInfo`;

/**
 * Sets the state of the processing component to a state displayable by the system.
 */
const EXPERIMENT_SETTINGS_PROCESSING_CONFIG_LOADED = `${EXPERIMENT_SETTINGS}/loadedProcessingConfig`;

/**
 * Sets error value in store if there is error in loading
 */
const EXPERIMENT_SETTINGS_PROCESSING_ERROR = `${EXPERIMENT_SETTINGS}/errorProcessing`;

/**
 * Updates the processing component's configuration.
 */
const EXPERIMENT_SETTINGS_NON_SAMPLE_FILTER_UPDATE = `${EXPERIMENT_SETTINGS}/updateNonSampleFilterSettings`;

/**
 * Updates the processing component's configuration.
 */
const EXPERIMENT_SETTINGS_SAMPLE_FILTER_UPDATE = `${EXPERIMENT_SETTINGS}/updateSampleFilterSettings`;

/**
 * Saves the current state of the configuration to DynamoDB.
 */
const EXPERIMENT_SETTINGS_PROCESSING_SAVE = `${EXPERIMENT_SETTINGS}/saveProcessing`;

/**
 * Signals that a pipeline run was initiated by the user.
 */
const EXPERIMENT_SETTINGS_QC_START = `${EXPERIMENT_SETTINGS}/startQC`;

/**
 * A QC step was enabled or disabled
 */
const EXPERIMENT_SETTINGS_SET_QC_STEP_ENABLED = `${EXPERIMENT_SETTINGS}/setQCStepEnabled`;

/**
 * The filter settings for one sample in one step were copied over to all other samples
 */
const EXPERIMENT_SETTINGS_COPY_SETTINGS_TO_ALL_SAMPLES = `${EXPERIMENT_SETTINGS}/copyFilterSettingsToAllSamples`;

const EXPERIMENT_SETTINGS_SET_SAMPLE_FILTER_SETTINGS_AUTO = `${EXPERIMENT_SETTINGS}/setSampleFilterSettingsAuto`;

/**
 * The filter settings for a filter were changed
 */
const EXPERIMENT_SETTINGS_ADD_CHANGED_QC_FILTER = `${EXPERIMENT_SETTINGS}/addChangedQCFilter`;

/**
 * The pending settings for filters were discarded
 */
const EXPERIMENT_SETTINGS_DISCARD_CHANGED_QC_FILTERS = `${EXPERIMENT_SETTINGS}/discardChangedQCFilters`;

/**
 * Received an update from qc for a sample in a step
 */
const EXPERIMENT_SETTINGS_UPDATE_SAMPLE_FROM_QC = `${EXPERIMENT_SETTINGS}/updateSampleProcessingSettingsFromQC`;

const EXPERIMENT_SETTINGS_PIPELINE_VERSION_UPDATED = `${EXPERIMENT_SETTINGS}/pipelineVersionUpdated`;

export {
  EXPERIMENT_SETTINGS_INFO_UPDATE,
  EXPERIMENT_SETTINGS_PROCESSING_CONFIG_LOADED,
  EXPERIMENT_SETTINGS_NON_SAMPLE_FILTER_UPDATE,
  EXPERIMENT_SETTINGS_SAMPLE_FILTER_UPDATE,
  EXPERIMENT_SETTINGS_PROCESSING_SAVE,
  EXPERIMENT_SETTINGS_PROCESSING_ERROR,
  EXPERIMENT_SETTINGS_QC_START,
  EXPERIMENT_SETTINGS_SET_QC_STEP_ENABLED,
  EXPERIMENT_SETTINGS_COPY_SETTINGS_TO_ALL_SAMPLES,
  EXPERIMENT_SETTINGS_SET_SAMPLE_FILTER_SETTINGS_AUTO,
  EXPERIMENT_SETTINGS_ADD_CHANGED_QC_FILTER,
  EXPERIMENT_SETTINGS_DISCARD_CHANGED_QC_FILTERS,
  EXPERIMENT_SETTINGS_UPDATE_SAMPLE_FROM_QC,
  EXPERIMENT_SETTINGS_PIPELINE_VERSION_UPDATED,
};
