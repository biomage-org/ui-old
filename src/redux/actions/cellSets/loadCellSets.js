import fetchAPI from 'utils/http/fetchAPI';
import handleError from 'utils/http/handleError';
import {
  CELL_SETS_LOADED, CELL_SETS_LOADING, CELL_SETS_ERROR,
} from 'redux/actionTypes/cellSets';
import endUserMessages from 'utils/endUserMessages';

import config from 'config';
import { api } from 'utils/constants';

const loadCellSets = (experimentId, forceReload = false) => async (dispatch, getState) => {
  const {
    loading, error, updatingClustering,
  } = getState().cellSets;

  if (!forceReload && ((!loading && !error) || updatingClustering)) {
    return null;
  }

  if (error) {
    dispatch({
      type: CELL_SETS_LOADING,
    });
  }

  let url;

  if (config.currentApiVersion === api.V1) {
    url = `/v1/experiments/${experimentId}/cellSets`;
  } else if (config.currentApiVersion === api.V2) {
    url = `/v2/experiments/${experimentId}/cellSets`;
  }

  try {
    const data = await fetchAPI(url);

    dispatch({
      type: CELL_SETS_LOADED,
      payload: {
        experimentId,
        data: data.cellSets,
      },
    });
  } catch (e) {
    const errorMessage = handleError(e, endUserMessages.ERROR_FETCHING_CELL_SETS);

    dispatch({
      type: CELL_SETS_ERROR,
      payload: { error: errorMessage },
    });
  }
};

export default loadCellSets;
