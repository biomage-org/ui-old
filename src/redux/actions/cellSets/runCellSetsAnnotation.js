import {
  CELL_SETS_ERROR, CELL_SETS_CLUSTERING_UPDATING,
} from 'redux/actionTypes/cellSets';

import getTimeoutForWorkerTask from 'utils/getTimeoutForWorkerTask';
import fetchWork from 'utils/work/fetchWork';

const runCellSetsAnnotation = (experimentId, species, tissue) => async (dispatch, getState) => {
  const { error, updatingClustering, loading } = getState().cellSets;

  if ((loading && updatingClustering) || error) return;

  const body = {
    name: 'ScTypeAnnotate',
    species,
    tissue,
  };

  dispatch({
    type: CELL_SETS_CLUSTERING_UPDATING,
  });

  const timeout = getTimeoutForWorkerTask(getState(), 'ClusterCells');

  try {
    await fetchWork(
      experimentId,
      body,
      getState,
      dispatch,
      {
        timeout,
        broadcast: true,
      },
    );
  } catch (e) {
    dispatch({
      type: CELL_SETS_ERROR,
      payload: {
        experimentId,
        error: e,
      },
    });
  }
};

export default runCellSetsAnnotation;
