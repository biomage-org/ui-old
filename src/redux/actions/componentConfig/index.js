import loadPlotConfig from 'redux/actions/componentConfig/loadPlotConfig';
import loadComponentConfig from 'redux/actions/componentConfig/loadComponentConfig';
import updatePlotConfig from 'redux/actions/componentConfig/updatePlotConfig';
import savePlotConfig from 'redux/actions/componentConfig/savePlotConfig';
import getDotPlot from 'redux/actions/componentConfig/getDotPlot';
import updatePlotData from 'redux/actions/componentConfig/updatePlotData';
import resetPlotConfig from 'redux/actions/componentConfig/resetPlotConfig';

export {
  loadPlotConfig,
  loadComponentConfig,
  updatePlotConfig,
  savePlotConfig,
  getDotPlot,
  updatePlotData,
  resetPlotConfig,
};
