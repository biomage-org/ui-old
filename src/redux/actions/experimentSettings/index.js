import saveProcessingSettings from './processingConfig/saveProcessingSettings';
import loadProcessingSettings from './processingConfig/loadProcessingSettings';
import loadedProcessingConfig from './processingConfig/loadedProcessingConfig';
import setQCStepEnabled from './processingConfig/setQCStepEnabled';
import copyFilterSettingsToAllSamples from './processingConfig/copyFilterSettingsToAllSamples';
import setSampleFilterSettingsAuto from './processingConfig/setSampleFilterSettingsAuto';
import updateFilterSettings from './processingConfig/updateFilterSettings';
import addChangedQCFilter from './processingConfig/addChangedQCFilter';
import discardChangedQCFilters from './processingConfig/discardChangedQCFilters';
import updateProcessingSettingsFromQC from './processingConfig/updateProcessingSettingsFromQC';
import updateExperimentInfo from './updateExperimentInfo';
import updatePipelineVersion from './updatePipelineVersion';

export {
  loadProcessingSettings,
  loadedProcessingConfig,
  saveProcessingSettings,
  updateExperimentInfo,
  setQCStepEnabled,
  copyFilterSettingsToAllSamples,
  setSampleFilterSettingsAuto,
  addChangedQCFilter,
  discardChangedQCFilters,
  updateFilterSettings,
  updateProcessingSettingsFromQC,
  updatePipelineVersion,
};
