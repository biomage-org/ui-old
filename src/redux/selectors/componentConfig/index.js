import getPlotConfigs from './getPlotConfigs';

/* eslint-disable-next-line import/prefer-default-export */
export { getPlotConfigs };
