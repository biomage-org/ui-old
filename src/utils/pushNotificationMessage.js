import { message } from 'antd';

const pushNotificationMessage = (type, text, duration = 4) => {
  switch (type) {
    case 'success':
      message.success(text, duration);
      break;
    case 'error':
      message.error(text, duration);
      break;
    case 'info':
      message.info(text, duration);
      break;
    case 'warning':
    case 'warn':
      message.warn(text, duration);
      break;
    case 'loading':
      message.loading(text, duration);
      break;
    default:
      message.info(text, duration);
      break;
  }
};

export default pushNotificationMessage;
