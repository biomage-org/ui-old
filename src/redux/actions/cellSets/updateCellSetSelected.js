import { CELL_SETS_SET_SELECTED } from '../../actionTypes/cellSets';

const updateCellSetSelected = (keys) => async (dispatch, getState) => {
  const {
    loading, error,
  } = getState().cellSets;

  if (loading || error) {
    return null;
  }

  await dispatch({
    type: CELL_SETS_SET_SELECTED,
    payload: {
      keys,
    },
  });
};

export default updateCellSetSelected;
