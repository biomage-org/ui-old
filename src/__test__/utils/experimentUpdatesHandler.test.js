import experimentUpdatesHandler, { updateTypes } from 'utils/experimentUpdatesHandler';
import fake from '__test__/test-utils/constants';
import pushNotificationMessage from 'utils/pushNotificationMessage';

import { updateCellSetsClustering, loadCellSets } from 'redux/actions/cellSets';
import { updateProcessingSettingsFromQC, loadedProcessingConfig, updatePipelineVersion } from 'redux/actions/experimentSettings';
import { updateBackendStatus } from 'redux/actions/backendStatus';
import { updatePlotData } from 'redux/actions/componentConfig';
import endUserMessages from 'utils/endUserMessages';

jest.mock('redux/actions/cellSets/updateCellSetsClustering');
jest.mock('redux/actions/experimentSettings', () => ({
  updateProcessingSettingsFromQC: jest.fn(),
  loadedProcessingConfig: jest.fn(),
  updatePipelineVersion: jest.fn(),
}));
jest.mock('redux/actions/backendStatus/updateBackendStatus');
jest.mock('redux/actions/componentConfig/updatePlotData');
jest.mock('redux/actions/cellSets/loadCellSets');

jest.mock('utils/pushNotificationMessage');
jest.spyOn(global.console, 'error');

const mockDispatch = jest.fn();

const triggerExperimentUpdate = (update) => {
  const actionCreator = experimentUpdatesHandler(mockDispatch);
  actionCreator(fake.EXPERIMENT_ID, update);
};

const mockError = {
  error: true,
  errorCode: 'MOCK_ERROR_CODE',
  userMessage: 'mockUserMessage',
};

const mockBackendStatusError = {
  pipeline: {
    status: 'ERROR',
  },
};

const mockQcUpdate = {
  type: updateTypes.QC,
  input: {
    taskName: 'mockTaskName',
    sampleUuid: 'mockSampleUuid',
  },
  output: {
    config: {
      processingConfigUpdate: {
        mockProcessingConfigUpdate: 'mockProcessingConfigUpdate',
      },
    },
    plotData: {
      mockPlotUuid: 'mockPlotData',
    },
  },
  response: {
    error: false,
  },
  status: {
    pipeline: {
      status: 'RUNNING',
    },
  },
  pipelineVersion: 2,
};

const mockGem2sUpdate = {
  type: updateTypes.GEM2S,
  item: {
    processingConfig: {
      mockProcessingConfig: 'mockProcessingConfig',
    },
  },
  status: {
    pipeline: {
      status: 'RUNNING',
    },
  },
};

const mockWorkResponseUpdate = {
  type: updateTypes.WORK_RESPONSE,
  request: {
    body: {
      name: 'WorkResponse',
    },
  },
  response: {
    error: false,
  },
};

describe('ExperimentUpdatesHandler', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Logs error if there is an error returned as an experiment update', () => {
    const mockUpdate = {
      ...mockQcUpdate,
      response: mockError,
      status: mockBackendStatusError,
    };

    triggerExperimentUpdate(mockUpdate);

    expect(global.console.error).toHaveBeenCalledTimes(1);
    expect(global.console.error).toMatchSnapshot();
  });

  it('Triggers properly for GEM2S updates ', () => {
    const mockUpdate = mockGem2sUpdate;

    triggerExperimentUpdate(mockUpdate);

    // Dispatch 1 - Update backend status
    expect(updateBackendStatus).toHaveBeenCalledTimes(1);
    const backendStatus = updateBackendStatus.mock.calls[0];
    expect(backendStatus).toMatchSnapshot();

    // Dispatch 2 - Loaded processingConfig
    const updateParams = loadedProcessingConfig.mock.calls[1];
    expect(updateParams).toMatchSnapshot();

    expect(mockDispatch).toHaveBeenCalledTimes(2);
    expect(pushNotificationMessage).not.toHaveBeenCalled();
  });

  it('Triggers properly for QC updates ', () => {
    const mockUpdate = mockQcUpdate;

    triggerExperimentUpdate(mockUpdate);

    // Dispatch 1 - Update backend status
    expect(updateBackendStatus).toHaveBeenCalledTimes(1);
    const backendStatus = updateBackendStatus.mock.calls[0];

    expect(backendStatus).toMatchSnapshot();

    // Dispatch 2 - update processin settings
    expect(updateProcessingSettingsFromQC).toHaveBeenCalledTimes(1);
    const updateParams = updateProcessingSettingsFromQC.mock.calls[0];
    expect(updateParams).toMatchSnapshot();

    // Dispatch 3 - update pipeline version
    expect(updatePipelineVersion).toHaveBeenCalledTimes(1);
    expect(updatePipelineVersion.mock.calls).toMatchSnapshot();

    // Dispatch 4 - update plot data
    expect(updatePlotData).toHaveBeenCalledTimes(1);
    const plotDataParams = updatePlotData.mock.calls[0];
    expect(plotDataParams).toMatchSnapshot();

    expect(mockDispatch).toHaveBeenCalledTimes(4);

    // Does not load cell sets
    expect(loadCellSets).not.toHaveBeenCalled();

    expect(pushNotificationMessage).not.toHaveBeenCalled();
  });

  it('Loads cell sets if QC pipeline completes ', () => {
    const mockUpdate = {
      ...mockQcUpdate,
      status: {
        pipeline: {
          status: 'SUCCEEDED',
        },
      },
    };

    triggerExperimentUpdate(mockUpdate);

    // Dispatch 1 - Update backend status
    expect(updateBackendStatus).toHaveBeenCalledTimes(1);
    const backendStatus = updateBackendStatus.mock.calls[0];
    expect(backendStatus).toMatchSnapshot();

    // Dispatch 2 - update plot data
    expect(updatePlotData).toHaveBeenCalledTimes(1);
    const plotDataParams = updatePlotData.mock.calls[0];
    expect(plotDataParams).toMatchSnapshot();

    // Dispatch 3 - update pipeline version
    expect(updatePipelineVersion).toHaveBeenCalledTimes(1);
    expect(updatePipelineVersion.mock.calls).toMatchSnapshot();

    // Dispatch 3 - load cellsets on pipeline finish
    expect(loadCellSets).toHaveBeenCalledTimes(1);
    const loadCellSetsParams = loadCellSets.mock.calls[0];
    expect(loadCellSetsParams).toMatchSnapshot();

    expect(mockDispatch).toHaveBeenCalledTimes(5);
    expect(pushNotificationMessage).not.toHaveBeenCalled();
  });

  it('Triggers properly for WorkResponse - ClusterCells updates ', () => {
    const mockUpdate = {
      ...mockWorkResponseUpdate,
      request: {
        body: {
          name: 'ClusterCells',
        },
      },
    };

    triggerExperimentUpdate(mockUpdate);

    // Dispatch 1 - Update cell set clustering
    expect(updateCellSetsClustering).toHaveBeenCalledTimes(1);

    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).toHaveBeenCalledWith(
      'success',
      endUserMessages.SUCCESS_CELL_SETS_RECLUSTERED,
    );
  });

  it('Triggers properly for WorkResponse - GetExpressionCellSets updates ', () => {
    const mockUpdate = {
      ...mockWorkResponseUpdate,
      request: {
        body: {
          name: 'GetExpressionCellSets',
        },
      },
    };

    triggerExperimentUpdate(mockUpdate);

    // Dispatch 1 - Load cell sets
    expect(loadCellSets).toHaveBeenCalledTimes(1);

    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).toHaveBeenCalledWith(
      'success',
      endUserMessages.SUCCESS_NEW_CLUSTER_CREATED,
    );
  });
});
