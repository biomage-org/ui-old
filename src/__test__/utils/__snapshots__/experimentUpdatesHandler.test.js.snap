// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExperimentUpdatesHandler Loads cell sets if QC pipeline completes  1`] = `
Array [
  "testae48e318dab9a1bd0bexperiment",
  Object {
    "pipeline": Object {
      "status": "SUCCEEDED",
    },
  },
]
`;

exports[`ExperimentUpdatesHandler Loads cell sets if QC pipeline completes  2`] = `
Array [
  "mockPlotUuid",
  "mockPlotData",
]
`;

exports[`ExperimentUpdatesHandler Loads cell sets if QC pipeline completes  3`] = `
Array [
  Array [
    "testae48e318dab9a1bd0bexperiment",
    2,
  ],
]
`;

exports[`ExperimentUpdatesHandler Loads cell sets if QC pipeline completes  4`] = `
Array [
  "testae48e318dab9a1bd0bexperiment",
  true,
]
`;

exports[`ExperimentUpdatesHandler Logs error if there is an error returned as an experiment update 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Experiment updates error:",
      Object {
        "input": Object {
          "sampleUuid": "mockSampleUuid",
          "taskName": "mockTaskName",
        },
        "output": Object {
          "config": Object {
            "processingConfigUpdate": Object {
              "mockProcessingConfigUpdate": "mockProcessingConfigUpdate",
            },
          },
          "plotData": Object {
            "mockPlotUuid": "mockPlotData",
          },
        },
        "pipelineVersion": 2,
        "response": Object {
          "error": true,
          "errorCode": "MOCK_ERROR_CODE",
          "userMessage": "mockUserMessage",
        },
        "status": Object {
          "pipeline": Object {
            "status": "ERROR",
          },
        },
        "type": "qc",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`ExperimentUpdatesHandler Triggers properly for GEM2S updates  1`] = `
Array [
  "testae48e318dab9a1bd0bexperiment",
  Object {
    "pipeline": Object {
      "status": "RUNNING",
    },
  },
]
`;

exports[`ExperimentUpdatesHandler Triggers properly for GEM2S updates  2`] = `undefined`;

exports[`ExperimentUpdatesHandler Triggers properly for QC updates  1`] = `
Array [
  "testae48e318dab9a1bd0bexperiment",
  Object {
    "pipeline": Object {
      "status": "RUNNING",
    },
  },
]
`;

exports[`ExperimentUpdatesHandler Triggers properly for QC updates  2`] = `
Array [
  "mockTaskName",
  Object {
    "processingConfigUpdate": Object {
      "mockProcessingConfigUpdate": "mockProcessingConfigUpdate",
    },
  },
  "mockSampleUuid",
  false,
]
`;

exports[`ExperimentUpdatesHandler Triggers properly for QC updates  3`] = `
Array [
  Array [
    "testae48e318dab9a1bd0bexperiment",
    2,
  ],
]
`;

exports[`ExperimentUpdatesHandler Triggers properly for QC updates  4`] = `
Array [
  "mockPlotUuid",
  "mockPlotData",
]
`;
